rfcs/20230418-constant-tensor-cache-capacity-control/README.md

```diff
@@ -0,0 +1,441 @@
+    size_t get_capacity();
+
+    size_t get_size() const;
+
+    /// @backend_id:
+    /// @backend_specific_key:
+    /// @size: The size the constant tensor
+    /// @value: A promised future constant tensor.
+    /// @return A valid future of cache hit, and an invalid one if cache miss.
+    value_t get_or_add(size_t backend_id, size_t backend_specific_key,
+            size_t size, const value_t &value);
+
+    void remove_if_exist(size_t backend_id, size_t backend_specific_key);
+};
+
+template<typename engine_kind>
+unified_constant_tensor_cache_t &get_unified_constant_tensor_cache_t();
+```
+
+The `set_capacity` and `get_capacity`are the
+implementations of the frontend capacity control APIs.
+
+The `set_capacity` must not be used by backends. The `get_capacity` can be used
+by backends to query information. Besides, the following interfaces are also
+provided to backends as a part of backend API:
+
+- `get_or_add`: this API is used to query a constant tensor from the cache. When
+  using this API, backend should pass in an empty `value` argument. If cache
+  hit, the returned `value_t` will be valid, and the passed empty `value` will
+  be ignored. If cache miss, the returned `value_t` will be invalid and the
+  behavior for empty `value` is different depending on whether the cache is full
+  or not. We will elaborate the cache miss situation later.
+- `remove_if_exist`: this API is used to remove a tensor from the cache. If the
+  removed tensor is currently being used, backend should hold it and make sure
+  it's alive.
+- `get_size`: this API is used to query the total size of all cached tensors.
+
+The `get_or_add` API behaviors when cache miss:
+
+If cache is not full, the empty `value` will be added to cache as a promise, and
+then the backend compute the constant tensor and set it to the empty `value` to
+fulfill the promise. In concurrent execution, if multiple threads query cache
+with same key, only one thread will cache miss and do the computation, other
+threads will cache hit and wait the promise to be fulfilled. This behavior is
+exactly same as the existing constant cache behaviors in DNNL backend.
+
+If cache is full, the empty `value` won't be added to cache and backend still
+need to compute the constant tensor.
+
+After a cached tensor is queried out by a backend, this tensor is co-owned by
+both unified cache and backend. Backend can hold the queried tensor handle
+temporarily. If the tensor is evicted from the unified cache during compiled
+partition execution, backend could still use the queried handle in current
+execution, but should release the tensor when the execution finished.
+
+To support concurrent execution, all interfaces in the cache class should be
+thread safe. It's guaranteed by library implementation.
+
+The cache class will be instantiated for each engine kind and the instances will
+have global lifecycle and will be destroyed when the library unloaded.
+
+## Open Questions
```

Comment:
And open questions better be removed till the next iteration, too.

Reply:
Removed, thanks! I will add this section somewhere in case of some future discussion. cc @TaoLv